### PR TITLE
Simplify ATen sparse semi-structured operators based on CUTLASS

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123742
* __->__ #123473

cc @ptrblck @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by alexsamardzic** [comment]
These updates remove `_sparse_semi_structured_linear` operator, and introduce `_sparse_semi_structured_(mm|addmm)` operators instead; namely, the former operator has a tensor as an optional argument, and it's not possible to register lowering in the Inductor for such an operator (see [here](https://github.com/pytorch/pytorch/blob/6890333e3dec55499d87d2193d03109f04b8779a/torch/_inductor/lowering.py#L288-L290)).

<details>
 <summary>Updated script to benchmark between cuSPARSELt and CUTLASS sparse semi-structured back-ends (also to compare results and benchmark vs. reference dense GEMM)</summary>

```Python
#! /usr/bin/env python

import random

import torch
from torch.testing import make_tensor
import torch.utils.benchmark as benchmark

from torch.sparse.semi_structured import (
    SparseSemiStructuredTensor,
    to_sparse_semi_structured,
)

SEMI_STRUCTURED_SUPPORTED_DTYPES = [
    torch.float16,
    torch.bfloat16,
    torch.float32,
    torch.int8,
]


def rand_sparse_semi_structured(r, c, dtype, device, choice=None):
    pattern = "2by4" if dtype != torch.float32 else "1by2"
    if pattern == "1by2":
        ksparse = 2
        choices = [[0, 1], [1, 0]]
    elif pattern == "2by4":
        ksparse = 4
        choices = [
            [1, 1, 0, 0],
            [1, 0, 1, 0],
            [1, 0, 0, 1],
            [0, 1, 1, 0],
            [0, 1, 0, 1],
            [0, 0, 1, 1],
        ]
    mask_entries = [choice or random.choice(choices) for i in range(r * c // ksparse)]
    mask = torch.tensor(mask_entries, dtype=torch.bool).view(r, c).to(device)
    dense = make_tensor(r, c, dtype=dtype, device=device)
    dense[dense == 0] = 1  # To prevent zeros except where mask applied.
    dense = dense.masked_fill(~mask, 0)
    return dense


def run_benchmark(m, n, k, device, dtype, rtol, atol, results):
    label = f"dense vs. cuSPARSELt vs. CUTLASS ({dtype})"
    sub_label = f"m:{m:5d} | k:{k:5d} | n:{n:5d}"

    dtypes_out = {
        torch.int8: torch.int32,
        torch.float16: torch.float16,
        torch.bfloat16: torch.bfloat16,
        torch.float32: torch.float32,#! /usr/bin/env python

import random

import torch
from torch.testing import make_tensor
import torch.utils.benchmark as benchmark

from torch.sparse.semi_structured import (
    SparseSemiStructuredTensor,
    to_sparse_semi_structured,
)

SEMI_STRUCTURED_SUPPORTED_DTYPES = [
    torch.float16,
    torch.bfloat16,
    torch.float32,
    torch.int8,
]


def rand_sparse_semi_structured(r, c, dtype, device, choice=None):
    pattern = "2by4" if dtype != torch.float32 else "1by2"
    if pattern == "1by2":
        ksparse = 2
        choices = [[0, 1], [1, 0]]
    elif pattern == "2by4":
        ksparse = 4
        choices = [
            [1, 1, 0, 0],
            [1, 0, 1, 0],
            [1, 0, 0, 1],
            [0, 1, 1, 0],
            [0, 1, 0, 1],
            [0, 0, 1, 1],
        ]
    mask_entries = [choice or random.choice(choices) for i in range(r * c // ksparse)]
    mask = torch.tensor(mask_entries, dtype=torch.bool).view(r, c).to(device)
    dense = make_tensor(r, c, dtype=dtype, device=device)
    dense[dense == 0] = 1  # To prevent zeros except where mask applied.
    dense = dense.masked_fill(~mask, 0)
    return dense


def run_benchmark(m, n, k, device, dtype, rtol, atol, results):
    label = f"dense vs. cuSPARSELt vs. CUTLASS ({dtype})"
    sub_label = f"m:{m:5d} | k:{k:5d} | n:{n:5d}"

    dtypes_out = {
        torch.int8: torch.int32,
        torch.float16: torch.float16,
        torch.bfloat16: torch.bfloat16,
        torch.float32: torch.float32,
    }
    dtype_out = dtypes_out[dtype]

    weight = rand_sparse_semi_structured(m, k, dtype, device)
    input = make_tensor((n, k), dtype=dtype, device=device).t()

    dtype_dense = torch.float32
    input_dense = input.to(dtype_dense)
    weight_dense = weight.to(dtype_dense)
    output0 = torch.mm(weight_dense, input_dense)

    if dtype.itemsize != 2:
        SparseSemiStructuredTensor._FORCE_CUTLASS = False
        compressed_cusparselt = to_sparse_semi_structured(weight)

        g = torch.cuda.CUDAGraph()
        with torch.cuda.graph(g):
            torch.mm(compressed_cusparselt, input)

        def mm_cusparselt():
            g.replay()

        output1 = mm_cusparselt()
        if dtype_out == dtype:
            torch.testing.assert_close(
                output1.to(dtype_dense), output0, rtol=rtol, atol=atol
        )

    SparseSemiStructuredTensor._FORCE_CUTLASS = True
    compressed_cutlass = to_sparse_semi_structured(weight)
    output2 = torch.mm(compressed_cutlass, input)
    if dtype_out == dtype:
        torch.testing.assert_close(output2.to(dtype_dense), output0, rtol=rtol, atol=atol)

    if dtype != torch.int8:
        measurement = benchmark.Timer(
            stmt="mm(mat1, mat2)",
            globals={
                "mm": torch.mm,
                "mat1": weight,
                "mat2": input,
            },
            label=label,
            sub_label=sub_label,
            description="dense latency",
        ).blocked_autorange()
        results.append(measurement)

    if dtype.itemsize != 2:
        measurement = benchmark.Timer(
            stmt="mm()",
            globals={
                "mm": mm_cusparselt(),
            },
            label=label,
            sub_label=sub_label,
            description="cuSPARSELt latency",
        ).blocked_autorange()
        results.append(measurement)
    
    measurement = benchmark.Timer(
        stmt="mm(mat1, mat2)",
        globals={
            "mm": torch.mm,
            "mat1": compressed_cutlass,
            "mat2": input,
        },
        label=label,
        sub_label=sub_label,
        description="CUTLASS latency",
    ).blocked_autorange()
    results.append(measurement)


if __name__ == "__main__":
    device = "cuda"

    shapes = [
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
        # llama shapes
        (4096, 11008, 64),
        (12288, 4096, 64),
        (22016, 4096, 64),
        (4096, 4096, 64),
        (65536, 4096, 64),
        (4096, 11008, 16),
        (12288, 4096, 16),
        (22016, 4096, 16),
        (4096, 4096, 16),
    ]

    torch.set_printoptions(
        precision=3,
        threshold=None,
        edgeitems=4,
        linewidth=460,
        profile=None,
        sci_mode=False,
    )

    for dtype in SEMI_STRUCTURED_SUPPORTED_DTYPES:
        results = []

        if dtype == torch.float32:
            # Inputs are converted to TF32 internally for sparse GEMM,
            # so make dense GEMM to do the same for matching results.
            orig = torch.backends.cuda.matmul.allow_tf32
            torch.backends.cuda.matmul.allow_tf32 = True

        rtol, atol = 1e-3, 1e-3
        if dtype == torch.float16 or dtype == torch.bfloat16:
            rtol, atol = 5e-3, 5e-2
        if dtype == torch.float32:
            rtol, atol = 1e-3, 5e-1

        for m, k, n in shapes:
            try:
                print(
                    f"m = {m:5d}, n = {n:5d}, k = {k:5d}, dtype = {dtype} ... ",
                    end="",
                )
                run_benchmark(m, n, k, device, dtype, rtol, atol, results)
                print("ok")
            except Exception:
                print("failed")
                continue

        if dtype == torch.float32:
            torch.backends.cuda.matmul.allow_tf32 = orig

        compare = benchmark.Compare(results)
        compare.colorize(rowwise=True)
        compare.print()

    }
    dtype_out = dtypes_out[dtype]

    weight = rand_sparse_semi_structured(m, k, dtype, device)
    input = make_tensor((n, k), dtype=dtype, device=device).t()

    dtype_dense = torch.float32
    input_dense = input.to(dtype_dense)
    weight_dense = weight.to(dtype_dense)
    output0 = torch.mm(weight_dense, input_dense)

    if dtype.itemsize != 2:
        SparseSemiStructuredTensor._FORCE_CUTLASS = False
        compressed_cusparselt = to_sparse_semi_structured(weight)

        g = torch.cuda.CUDAGraph()
        with torch.cuda.graph(g):
            torch.mm(compressed_cusparselt, input)

        def mm_cusparselt():
            g.replay()

        output1 = mm_cusparselt()
        if dtype_out == dtype:
            torch.testing.assert_close(
                output1.to(dtype_dense), output0, rtol=rtol, atol=atol
        )

    SparseSemiStructuredTensor._FORCE_CUTLASS = True
    compressed_cutlass = to_sparse_semi_structured(weight)
    output2 = torch.mm(compressed_cutlass, input)
    if dtype_out == dtype:
        torch.testing.assert_close(output2.to(dtype_dense), output0, rtol=rtol, atol=atol)

    if dtype != torch.int8:
        measurement = benchmark.Timer(
            stmt="mm(mat1, mat2)",
            globals={
                "mm": torch.mm,
                "mat1": weight,
                "mat2": input,
            },
            label=label,
            sub_label=sub_label,
            description="dense latency",
        ).blocked_autorange()
        results.append(measurement)

    if dtype.itemsize != 2:
        measurement = benchmark.Timer(
            stmt="mm()",
            globals={
                "mm": mm_cusparselt(),
            },
            label=label,
            sub_label=sub_label,
            description="cuSPARSELt latency",
        ).blocked_autorange()
        results.append(measurement)
    
    measurement = benchmark.Timer(
        stmt="mm(mat1, mat2)",
        globals={
            "mm": torch.mm,
            "mat1": compressed_cutlass,
            "mat2": input,
        },
        label=label,
        sub_label=sub_label,
        description="CUTLASS latency",
    ).blocked_autorange()
    results.append(measurement)


if __name__ == "__main__":
    device = "cuda"

    shapes = [
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
        # llama shapes
        (4096, 11008, 64),
        (12288, 4096, 64),
        (22016, 4096, 64),
        (4096, 4096, 64),
        (65536, 4096, 64),
        (4096, 11008, 16),
        (12288, 4096, 16),
        (22016, 4096, 16),
        (4096, 4096, 16),
    ]

    torch.set_printoptions(
        precision=3,
        threshold=None,
        edgeitems=4,
        linewidth=460,
        profile=None,
        sci_mode=False,
    )

    for dtype in SEMI_STRUCTURED_SUPPORTED_DTYPES:
        results = []

        if dtype == torch.float32:
            # Inputs are converted to TF32 internally for sparse GEMM,
            # so make dense GEMM to do the same for matching results.
            orig = torch.backends.cuda.matmul.allow_tf32
            torch.backends.cuda.matmul.allow_tf32 = True

        rtol, atol = 1e-3, 1e-3
        if dtype == torch.float16 or dtype == torch.bfloat16:
            rtol, atol = 5e-3, 5e-2
        if dtype == torch.float32:
            rtol, atol = 1e-3, 5e-1

        for m, k, n in shapes:
            try:
                print(
                    f"m = {m:5d}, n = {n:5d}, k = {k:5d}, dtype = {dtype} ... ",
                    end="",
                )
                run_benchmark(m, n, k, device, dtype, rtol, atol, results)
                print("ok")
            except Exception:
                print("failed")
                continue

        if dtype == torch.float32:
            torch.backends.cuda.matmul.allow_tf32 = orig

        compare = benchmark.Compare(results)
        compare.colorize(rowwise=True)
        compare.print()
```
</details>

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3342,9 +3342,13 @@
   dispatch:
     CUDA: _cslt_sparse_mm_search
 
-- func: _sparse_semi_structured_linear(Tensor input, Tensor weight, Tensor meta, *, Tensor? bias=None, str? activation=None, ScalarType? out_dtype=None) -> Tensor
```
[the remark this answers is not in the document]
Comment (alexsamardzic) [reply aten/src/ATen/native/native_functions.yaml:3345]: Reverted removal of `_sparse_semi_structured_linear` operator, and added deprecation warning.

**Review by cpuhrsch — APPROVED** [review]

**Comment by DanilBaibak** [comment]
Hi @alexsamardzic! Sorry, I need to revert your PR because it breaks the internal build. I suspect some internal CI flakiness. So please rebase your PR and merge it again. @cpuhrsch could you pls help if there are any questions.

**Comment by alexsamardzic** [comment]
Rebased on `viable/strict`, all the test pass, will merge again.

**Pull request closed without merging**